processing/src/main/java/io/druid/query/groupby/RowBasedColumnSelectorFactory.java

```diff
@@ -0,0 +1,313 @@
+import io.druid.math.expr.Parser;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.query.extraction.ExtractionFn;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.DimensionSelector;
+import io.druid.segment.FloatColumnSelector;
+import io.druid.segment.LongColumnSelector;
+import io.druid.segment.NumericColumnSelector;
+import io.druid.segment.ObjectColumnSelector;
+import io.druid.segment.column.Column;
+import io.druid.segment.column.ColumnCapabilities;
+import io.druid.segment.column.ColumnCapabilitiesImpl;
+import io.druid.segment.column.ValueType;
+import io.druid.segment.data.IndexedInts;
+import it.unimi.dsi.fastutil.ints.IntIterator;
+import it.unimi.dsi.fastutil.ints.IntIterators;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public class RowBasedColumnSelectorFactory implements ColumnSelectorFactory
+{
+  private final Supplier<? extends Row> row;
+  private final Map<String, ValueType> columnTypes;
+
+  public RowBasedColumnSelectorFactory(final Supplier<? extends Row> row, final Map<String, ValueType> columnTypes)
+  {
+    this.row = row;
+    this.columnTypes = columnTypes != null ? columnTypes : ImmutableMap.<String, ValueType>of();
+  }
+
+  public static RowBasedColumnSelectorFactory create(
+      final Supplier<? extends Row> row,
+      final Map<String, ValueType> columnTypes
+  )
+  {
+    return new RowBasedColumnSelectorFactory(row, columnTypes);
+  }
+
+  public static RowBasedColumnSelectorFactory create(
+      final ThreadLocal<? extends Row> row,
+      final Map<String, ValueType> columnTypes
+  )
+  {
+    return new RowBasedColumnSelectorFactory(
+        new Supplier<Row>()
+        {
+          @Override
+          public Row get()
+          {
+            return row.get();
+          }
+        },
+        columnTypes
+    );
+  }
+
+  // This dimension selector does not have an associated lookup dictionary, which means lookup can only be done
+  // on the same row. This dimension selector is used for applying the extraction function on dimension, which
+  // requires a DimensionSelector implementation
+  @Override
+  public DimensionSelector makeDimensionSelector(DimensionSpec dimensionSpec)
+  {
+    return dimensionSpec.decorate(makeDimensionSelectorUndecorated(dimensionSpec));
+  }
+
+  private DimensionSelector makeDimensionSelectorUndecorated(DimensionSpec dimensionSpec)
+  {
+    final String dimension = dimensionSpec.getDimension();
+    final ExtractionFn extractionFn = dimensionSpec.getExtractionFn();
+    final boolean isTimeColumn = dimensionSpec.getDimension().equals(Column.TIME_COLUMN_NAME);
+
+    if (isTimeColumn && extractionFn == null) {
+      throw new UnsupportedOperationException("time dimension must provide an extraction function");
+    }
+
+    return new DimensionSelector()
+    {
+      @Override
+      public IndexedInts getRow()
+      {
+        final int dimensionValuesSize;
+
+        if (isTimeColumn) {
```

Comment:
Sure I can split these. But, the timeColumn dimension selector does always return a row with `[0]` but `lookupName` does something different every time, so it can't be a singleton.